# HG changeset patch
# User Jordon Moss <mossj32@gmail.com>
# Date 1745200432 10800
#      Sun Apr 20 22:53:52 2025 -0300
# Node ID c68134624e1b531ea8565bb254af52d8afba3b4a
# Parent  31a8ee9802826d63f67220f06a08e59dbc060143
# EXP-Topic strikers_stuff
Add +LOCKEDTOCEILING and +LOCKEDTOFLOOR flags.

diff -r 31a8ee980282 -r c68134624e1b docs/zandronum-history.txt
--- a/docs/zandronum-history.txt	Fri Apr 18 08:39:01 2025 -0300
+++ b/docs/zandronum-history.txt	Sun Apr 20 22:53:52 2025 -0300
@@ -18,6 +18,7 @@
 *+	- Added +DECALUSESTRANSLATION, allowing an actor to pass its translation down to its decal. Useful to change decal color on the fly, or to pass the color of blood particles down to the decal. Pal index 255 is used set the color for shaded decals. [StrikerTheHedgefox]
 *+	- Added +DECALUSESSPRITESCALE, which multiplies an actor's decal scale by the sprite scale. [StrikerTheHedgefox]
 *+	- Added +NEVERADJUSTSPRITECLIPPING, which sets an actor to never adjust the clipping of its sprites to the floor in OpenGL, regardless of the gl_spriteclip setting. [StrikerTheHedgefox]
+*+	- Added +LOCKEDTOCEILING and +LOCKEDTOFLOOR, which locks an actor to the ceiling or floor, even if the actor itself isn't moving. (Unlike CEILINGHUGGER) [StrikerTheHedgefox]
 *+	- FakeInventory actors can now make use of the "Activation" thing property, vastly increasing its utility. Allows the item to target the player collecting the item and vice/versa, or for the item itself to be the activator. [StrikerTheHedgefox]
 *-	- Send inventory that inherits from CustomInventory and Inventory on full update. Fixes issues where weapon checks for inventory fails for late-joining clients, or after a map change. [StrikerTheHedgefox]
 +	- Made RandomSpawners pass AMBUSH, FRIENDLY, STEALTH, DORMANT, and STANDSTILL flags down to the actor if set via scripts. [StrikerTheHedgefox]
diff -r 31a8ee980282 -r c68134624e1b src/actor.h
--- a/src/actor.h	Fri Apr 18 08:39:01 2025 -0300
+++ b/src/actor.h	Sun Apr 20 22:53:52 2025 -0300
@@ -493,6 +493,8 @@
 	HFXF_DECALUSESTRANSLATION		= (1 << 0), // Decal will use actor's translation, or use the translation's pal index 255 color as the basis of its shade if untranslatable.
 	HFXF_DECALUSESSPRITESCALE		= (1 << 1), // Decal's scale will be multiplied by actor's sprite scale.
 	HFXF_NEVERADJUSTSPRITECLIPPING	= (1 << 2), // Never adjust clipping of this actor's sprites in OpenGL, regardless of the gl_spriteclip setting.
+	HFXF_LOCKEDTOCEILING			= (1 << 3), // Locks this actor to the ceiling, even if the actor isn't moving or isn't a missile. (Unlike CEILINGHUGGER)
+	HFXF_LOCKEDTOFLOOR				= (1 << 4), // Locks this actor to the floor, even if the actor isn't moving or isn't a missile. (Unlike FLOORHUGGER).
 
 // --- mobj.renderflags ---
 
diff -r 31a8ee980282 -r c68134624e1b src/p_mobj.cpp
--- a/src/p_mobj.cpp	Fri Apr 18 08:39:01 2025 -0300
+++ b/src/p_mobj.cpp	Sun Apr 20 22:53:52 2025 -0300
@@ -4077,6 +4077,14 @@
 		y += vely;
 		z += velz;
 		LinkToWorld ();
+
+		if (this->HFXFlags & (HFXF_LOCKEDTOCEILING | HFXF_LOCKEDTOFLOOR))
+		{
+			floorz = this->Sector->floorplane.ZatPoint(x, y);
+			ceilingz = this->Sector->ceilingplane.ZatPoint(x, y);
+
+			z = (this->HFXFlags & HFXF_LOCKEDTOCEILING) ? ceilingz : floorz;
+		}
 	}
 	else
 	{
@@ -4437,7 +4445,15 @@
 			}
 
 		}
-		if (velz || BlockingMobj || z != floorz)
+
+		if (this->HFXFlags & (HFXF_LOCKEDTOCEILING | HFXF_LOCKEDTOFLOOR))
+		{
+			floorz = this->Sector->floorplane.ZatPoint(x, y);
+			ceilingz = this->Sector->ceilingplane.ZatPoint(x, y);
+
+			z = (this->HFXFlags & HFXF_LOCKEDTOCEILING) ? ceilingz : floorz;
+		}
+		else if (velz || BlockingMobj || z != floorz)
 		{	// Handle Z velocity and gravity
 			if (((flags2 & MF2_PASSMOBJ) || (flags & MF_SPECIAL)) && !(i_compatflags & COMPATF_NO_PASSMOBJ))
 			{
@@ -4540,6 +4556,7 @@
 			if (!PoisonDurationReceived) PoisonDamageReceived = 0;
 		}
 	}
+
 	// [BC] Don't tick states while predicting.
 	if ( CLIENT_PREDICT_IsPredicting( ))
 		return;
diff -r 31a8ee980282 -r c68134624e1b src/thingdef/thingdef_data.cpp
--- a/src/thingdef/thingdef_data.cpp	Fri Apr 18 08:39:01 2025 -0300
+++ b/src/thingdef/thingdef_data.cpp	Sun Apr 20 22:53:52 2025 -0300
@@ -283,6 +283,8 @@
 	DEFINE_FLAG(HFXF, DECALUSESTRANSLATION, AActor, HFXFlags),
 	DEFINE_FLAG(HFXF, DECALUSESSPRITESCALE, AActor, HFXFlags),
 	DEFINE_FLAG(HFXF, NEVERADJUSTSPRITECLIPPING, AActor, HFXFlags),
+	DEFINE_FLAG(HFXF, LOCKEDTOCEILING, AActor, HFXFlags),
+	DEFINE_FLAG(HFXF, LOCKEDTOFLOOR, AActor, HFXFlags),
 
 	// [BB] New DECORATE network related flag defines here.
 	DEFINE_FLAG(NETFL, NONETID, AActor, NetworkFlags),
